File: bindings/java/src/test/java/org/indunet/robot/bus/TestBus.java
package org.indunet.robot.bus;

/** Ephemeral TCP broker + matching node options (mirrors C++ {@code TestBus}). */
final class TestBus implements AutoCloseable {
    final Broker broker;
    final String messageXsub;
    final String messageXpub;
    final String serviceFrontend;
    final String serviceBackend;
    final String actionFrontend;
    final String actionBackend;
    final String apiListen;

    private TestBus(
            Broker broker,
            String messageXsub,
            String messageXpub,
            String serviceFrontend,
            String serviceBackend,
            String actionFrontend,
            String actionBackend,
            String apiListen) {
        this.broker = broker;
        this.messageXsub = messageXsub;
        this.messageXpub = messageXpub;
        this.serviceFrontend = serviceFrontend;
        this.serviceBackend = serviceBackend;
        this.actionFrontend = actionFrontend;
        this.actionBackend = actionBackend;
        this.apiListen = apiListen;
    }

    static TestBus start() {
        // Bind :0 so the OS assigns ports at broker start (avoids freePort TOCTOU).
        BrokerOptions opts =
                new BrokerOptions(
                        "tcp://127.0.0.1:0",
                        "tcp://127.0.0.1:0",
                        "tcp://127.0.0.1:0",
                        "tcp://127.0.0.1:0",
                        "tcp://127.0.0.1:0",
                        "tcp://127.0.0.1:0",
                        "127.0.0.1:0",
                        null,
                        true,
                        true);
        Broker broker = new Broker(opts);
        return new TestBus(
                broker,
                broker.messageXsubBind(),
                broker.messageXpubBind(),
                broker.serviceFrontendBind(),
                broker.serviceBackendBind(),
                broker.actionFrontendBind(),
                broker.actionBackendBind(),
                broker.apiListen());
    }

    String wsUrl() {
        return "http://" + apiListen;
    }

    Node makeNode(String name) {
        return new Node(
                name,
                new NodeOptions(
                        null,
                        "tcp",
                        null,
                        messageXsub,
                        messageXpub,
                        serviceFrontend,
                        serviceBackend,
                        actionBackend,
                        actionFrontend));
    }

    @Override
    public void close() {
        broker.stop();
        broker.close();
    }
}

File: bindings/android/src/test/kotlin/org/indunet/robot/bus/TestBus.kt
package org.indunet.robot.bus

/** Ephemeral TCP broker + matching node options (mirrors Java / C++ TestBus). */
internal class TestBus private constructor(
    val broker: Broker,
    val messageXsub: String,
    val messageXpub: String,
    val serviceFrontend: String,
    val serviceBackend: String,
    val actionFrontend: String,
    val actionBackend: String,
    val apiListen: String,
) : AutoCloseable {
    fun wsUrl(): String = "http://$apiListen"

    fun makeNode(name: String): Node =
        Node(
            name,
            NodeOptions(
                host = null,
                transport = "tcp",
                wsUrl = null,
                messageXsub = messageXsub,
                messageXpub = messageXpub,
                serviceFrontend = serviceFrontend,
                serviceBackend = serviceBackend,
                actionBackend = actionBackend,
                actionFrontend = actionFrontend,
            ),
        )

    override fun close() {
        broker.stop()
        broker.close()
    }

    companion object {
        @JvmStatic
        fun start(): TestBus {
            // Bind :0 so the OS assigns ports at broker start (avoids freePort TOCTOU).
            val opts =
                BrokerOptions(
                    messageXsubBind = "tcp://127.0.0.1:0",
                    messageXpubBind = "tcp://127.0.0.1:0",
                    serviceFrontendBind = "tcp://127.0.0.1:0",
                    serviceBackendBind = "tcp://127.0.0.1:0",
                    actionFrontendBind = "tcp://127.0.0.1:0",
                    actionBackendBind = "tcp://127.0.0.1:0",
                    apiListen = "127.0.0.1:0",
                    consoleListen = null,
                    tcpOnly = true,
                    noConsole = true,
                )
            val broker = Broker(opts)
            return TestBus(
                broker,
                broker.messageXsubBind(),
                broker.messageXpubBind(),
                broker.serviceFrontendBind(),
                broker.serviceBackendBind(),
                broker.actionFrontendBind(),
                broker.actionBackendBind(),
                broker.apiListen(),
            )
        }
    }
}
